### DISCUSSION TITLE
Potentiometer as an encoder

### DISCUSSION THREAD

**Discussion by gringow831** [tttapa/Control-Surface#885] [General]
Is there a way to use a potentiometer as an encoder?  I know I could use pbabsoluteencoder for an encoder as a pb but wondering how could I control absoluteencoder with a p potentiometer giving it 0% to 100%

**Comment by tttapa** [comment]
Like `PBPotentiometer`? See https://tttapa.github.io/Control-Surface-doc/Doxygen/d7/dcd/group__MIDIOutputElements.html for the full list of available options.

**Reply by gringow831** [reply]
Yes, I am trying to control an encoder with a pbpotentiomer. Using an analog pot

**Reply by gringow831** [reply]
I believe it’s what the V_POT does using a Potentiometer rather than encoder

**Reply by tttapa** [reply]
MCU V-Pots are always relative, while a potentiometer is by definition absolute.

**Reply by gringow831** [reply]
That’s why I was wondering if there was a way to specify an absolute position like with a potentiometer or in the case of a Pbabsoluteencoder

**Reply by tttapa** [reply]
What do you mean by "specify" an absolute position? A potentiometer has an absolute position, you can only "specify" it by turning the knob. An encoder does not have an absolute position. PBAbsoluteEncoder maintains an absolute position in software, that you can access or change using the getters and setters that the class provides.

**Reply by gringow831** [reply]
For example controlling the gain input with a digidesign c24 it’s absolute and I believe they used a rotary potentiometer rather than an encoder for it. It controls the MCU V_pot on pro tools for when using it with their preamps and then linear faders for the volume. I’m trying to make a similar one

**Reply by gringow831** [reply]
Another example is for example the with panning and the v ring and using an encoder. As it shows an absolute position from left to right going around. I was wondering if I could use a potentiometer for that too rather than an encoder? So I can have only a potentiometer for all

**Reply by gringow831** [reply]
I found this but wouldn’t know how to apply it so for example I could use V_POT_1 
https://community.element14.com/products/arduino/f/forum/30059/potentiometer-instead-of-rotary-encoder

I understand a por has 1 wiper and encoders 2 pins plus + and GND and sw. is there a way to set the beginning and end parameter of the potentiometer like on an absolute encoder?

**Reply by gringow831** [reply]
I guess the way to explain what I am trying to do is to use a potentiometer as I haven't found encoders with 270º mechanical angle and dented-less.

**Reply by tttapa** [reply]
Encoders send relative signals. If you turn it one click to the left, it sends -1, if you turn it one click to the right, it sends +1.  
Potentiometers send absolute signals. If the shaft is rotated from 0° to 270°, it sends values between 0 and 1023 (or 0 and 16383).  

Within the limited 270° range of the potentiometer, you can turn the absolute position into relative deltas in software. However, when you hit the boundaries of the potentiometer travel, that no longer works. Since there is no way to keep the physical knob position in sync with the DAW's control value, there's no way to avoid this issue.

Here's some code that turns an absolute potentiometer position into relative CC messages if you want to experiment with it yourself, but I'm afraid you'll find that it is not useful:
```cpp
#include <Control_Surface.h> // Include the Control Surface library

USBDebugMIDI_Interface midi;

// Potentiometer that pretends to be a quadrature encoder
struct FakeEncoder {
    FakeEncoder(pin_t analogPin) : potentiometer(analogPin) {}
    FilteredAnalog<10> potentiometer;
    void begin() { potentiometer.setupADC(); }
    analog_t read() { return potentiometer.update(), potentiometer.getValue(); }
};

// Class that sends relative CC messages based on the position of a FakeEncoder
struct FakeCCEncoder : GenericMIDIRotaryEncoder<FakeEncoder, RelativeCCSender> {
    FakeCCEncoder(FakeEncoder &&encoder, MIDIAddress address,
                  int16_t speedMultiply = 1, uint8_t pulsesPerStep = 1)
        : GenericMIDIRotaryEncoder<FakeEncoder, RelativeCCSender>(
              std::move(encoder), address, speedMultiply, pulsesPerStep, {}) {}
};

FakeCCEncoder enc {A0, MCU::V_POT_1};

void setup() {
  RelativeCCSender::setMode(relativeCCmode::MACKIE_CONTROL_RELATIVE);
  Control_Surface.begin();
}

void loop() {
  Control_Surface.loop();
}
```

**Reply by gringow831** [reply]
Thank you so much Pieter, that explains a lot actually and I see what you did there. It's very useful and will try experimenting too.

**Reply by gringow831** [reply]
actually, it acts almost in the way I want it and I think I will need to map the dead zone

**Comment by gringow831** [comment]
First of all thank you so much for your help.

I have a few questions that you might be able to point me in the right direction. 
I have tried mapping the dead zones  [using this method](https://cdn.arduino.cc/reference/en/language/functions/math/map/?_gl=1*1wp8049*_ga*MTA0MjYzMjQ5Ni4xNjc1NDgyMTkx*_ga_NEXN8H46L5*MTY3NTgwOTA4OS4xMC4xLjE2NzU4MTYyNjkuMC4wLjA.) as when I use enc.map(mappingFunction) it returns the following error.

I assume the "has no member named map" it's because  it's not part of the library?

```
fake_encoder: In function 'void setup()':
fake_encoder:43: error: 'struct FakeCCEncoder' has no member named 'map'
 enc.map(mappingFunction);
     ^
'struct FakeCCEncoder' has no member named 'map'
```

To be able to Bank or ManyAddresses this I guess I will need to modify the library or is there a way without needing to create hpp and cpp files?

**Reply by tttapa** [reply]
> I assume the "has no member named map" it's because it's not part of the library?

No, it's just because the FakeCCEncoder class doesn't define such a member function. 

`map()` is a member function of the `FilteredAnalog` class and of the potentiometer classes such as `CCPotentiometer` and `PBPotentiometer`. 

If you want to apply a mapping function in your code, you should apply it to the `potentiometer` member of `FakeEncoder`. You could apply it in the `FakeEncoder` constructor for instance:

```cpp
// Potentiometer that pretends to be a quadrature encoder
struct FakeEncoder {
    FakeEncoder(pin_t analogPin) : potentiometer(analogPin) {
       potentiometer.map(yourMappingFunction);
    }
    FilteredAnalog<10> potentiometer;
    void begin() { potentiometer.setupADC(); }
    analog_t read() { return potentiometer.update(), potentiometer.getValue(); }
};
```

> To be able to Bank or ManyAddresses this I guess I will need to modify the library or is there a way without needing to create hpp and cpp files?

No, there's no need to edit the library, you can just write the code in your main sketch.

To make it bankable, simply replace `GenericMIDIRotaryEncoder<FakeEncoder, RelativeCCSender>` by `Bankable::GenericMIDIRotaryEncoder<FakeEncoder, Bankable::SingleAddress, RelativeCCSender>`, and edit the constructor to accept an additional input of type `OutputBankConfig<>`, e.g.

```cpp
namespace CS { namespace Bankable {

// Class that sends relative CC messages based on the position of a FakeEncoder
struct FakeCCEncoder : GenericMIDIRotaryEncoder<FakeEncoder, SingleAddress, RelativeCCSender> {
    FakeCCEncoder(OutputBankConfig<> config, FakeEncoder &&encoder, MIDIAddress address,
                  int16_t speedMultiply = 1, uint8_t pulsesPerStep = 1)
        : GenericMIDIRotaryEncoder<FakeEncoder, SingleAddress, RelativeCCSender>(
              {config, address}, std::move(encoder), speedMultiply,
              pulsesPerStep, {}) {}
};

}} // namespace CS::Bankable

Bankable::FakeCCEncoder enc {bank, A0, MCU::V_POT_1};
```
For ManyAddresses, it's the same thing, but with `ManyAddresses` instead of `SingleAddress`, and you need to specify the number of banks as a template parameter:
```cpp
namespace CS { namespace Bankable { namespace ManyAddresses {

// Class that sends relative CC messages based on the position of a FakeEncoder
template <setting_t NumBanks>
struct FakeCCEncoder : GenericMIDIRotaryEncoder<FakeEncoder, ManyAddresses<NumBanks>, RelativeCCSender> {
    FakeCCEncoder(const Bank<NumBanks> &bank, FakeEncoder &&encoder, const Array<MIDIAddress, NumBanks> &addresses,
                  int16_t speedMultiply = 1, uint8_t pulsesPerStep = 1)
        : GenericMIDIRotaryEncoder<FakeEncoder, ManyAddresses<NumBanks>, RelativeCCSender>(
              {bank, addresses}, std::move(encoder), speedMultiply,
              pulsesPerStep, {}) {}
};

}}} // namespace CS::Bankable::ManyAddresses

Bankable::ManyAddresses::FakeCCEncoder<4> enc {bank, A0, {MCU::V_POT_1, MCU::V_POT_8, MCU::V_POT_3, MCU::V_POT_5}};
```

As a side note, you can also play with the resolution of the `FilteredAnalog`, and with the `speedMultiply` and `pulsesPerStep` parameters to change the range and speed of your V-Pot.

**Reply by gringow831** [reply]
Works perfect! I tried for a while doing so and did't get any close to it today! 
Thank you

**Reply by gringow831** [reply]
Hi Pieter, is there a way to make the potentiometer used to remember it's last position as with the SmartPotentiometer?

**Reply by gringow831** [reply]
I have been looking at the SmartMIDIFilteredAnalog for a while but I don't seem to find the key to add it to the example in order to read last position of the Potentiometer when I change banks. 
https://github.com/tttapa/Control-Surface/issues/328#issuecomment-726399145 "You could make PBAbsoluteEncoder smarter by listening to the position feedback from the DAW. If you want to go that route, let me know and I'll provide some more information."

Not entirely sure if the smart bit needs to be the FakeCCEncoder used or the FakeEncoder AKA the read of the Potentiometer although I have tried with both parts
